### PR TITLE
Refactor object part stream

### PR DESCRIPTION
**Pull request by monthonk**
Various refactorings, including a new config type for object part stream task, introducing structs for part composers, consolidating error handling flow in request reader and part composer.

This is a part of https://github.com/awslabs/mountpoint-s3/pull/926.

## Does this change impact existing behavior?

No

## Does this change need a changelog entry in any of the crates?

No, changes are in mountpoint-s3 crate but it does not change any behavior.

---

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and I agree to the terms of the [Developer Certificate of Origin (DCO)](https://developercertificate.org/).

### CODE REVIEW

File: mountpoint-s3/src/prefetch/part_stream.rs [hunk 1]
```diff
@@ -188,18 +191,21 @@ where
...
+                        preferred_part_size: config.preferred_part_size,
+                    };
+
                     let request_reader_future =
-                        read_from_request(body_sender, client, bucket, object_id.clone(), request_range.into());
-                    let part_composer_future =
-                        compose_parts(body_receiver, part_queue_producer, object_id, preferred_part_size);
+                        try_read_from_request(body_sender, client, bucket, config.object_id, request_range.into());
+                    let part_composer_future = part_composer.try_compose_parts(body_receiver);
                     join!(request_reader_future, part_composer_future);
```
Comment (passaro) [inline mountpoint-s3/src/prefetch/part_stream.rs:209]: This `join` over two futures communicating over a channel seems quite convoluted. Have we considered modelling the two components as stream transformer and consumer instead?

Comment (passaro) [reply mountpoint-s3/src/prefetch/part_stream.rs:209]: This is not specific to your changes in this PR, though, so we can also discuss it separately.

Comment (passaro) [reply mountpoint-s3/src/prefetch/part_stream.rs:209]: Here is a throw-away example - just for `read_from_request` - which uses `async_stream`: https://github.com/awslabs/mountpoint-s3/compare/main...passaro:mountpoint-s3:async-stream

Comment (monthonk) [reply mountpoint-s3/src/prefetch/part_stream.rs:209]: We haven't considered this option yet. I will have a look at your example, but probably make a separate PR for it if we want to change.

File: mountpoint-s3/src/prefetch/part_stream.rs [hunk 2]
```diff
@@ -210,100 +216,95 @@ where
...
-                    curr_offset += part.len() as u64;
-                    part_queue_producer.push(Ok(part));
-                }
-            }
-            Some(Err(err)) => {
-                part_queue_producer.push(Err(err));
-                break;
-            }
-            None => break,
+pub async fn read_from_request<Client: ObjectClient>(
```
[the remark this answers is not in the document]
Comment (monthonk) [reply mountpoint-s3/src/prefetch/part_stream.rs:281]: Yes, this doesn't have to be public.

File: mountpoint-s3/src/prefetch/part_stream.rs (hunk 1 above)
Comment (monthonk) [reply mountpoint-s3/src/prefetch/part_stream.rs:209]: Did try `async_stream` and it was really neat! I would make it a followup PR after this one.

**Review by passaro — APPROVED** [review]
LGTM!

**Pull request merged**